Track validation errors and registration submitted events

yuki24 opened the pull request:
addresses https://artsyproduct.atlassian.net/browse/AUCT-465

It adds the `<OnSubmitValidationError>` component to rack validation errors from Formik.

@dblandin @erikdstock I'm not sure if I'm using it right, but let me know if there's anything I should update.

## Tracking validation errors

<img width="2032" alt="Screen Shot 2019-11-21 at 2 27 59 PM" src="https://user-images.githubusercontent.com/386234/69373448-cf7af100-0c71-11ea-8158-13395f961126.png">

## Tracking `registration submitted`

<img width="2032" alt="Screen Shot 2019-11-21 at 3 08 32 PM" src="https://user-images.githubusercontent.com/386234/69373460-d3a70e80-0c71-11ea-81cb-01af12232235.png">

dleve123 commented on src/Apps/Auction/Routes/__tests__/ConfirmBid.test.tsx:
```diff
@@ -394,8 +394,17 @@ describe("Routes/ConfirmBid", () => {
-      expect(mockPostEvent).toHaveBeenCalledTimes(1)
-      expect(mockPostEvent).toHaveBeenCalledWith({
+      expect(mockPostEvent).toHaveBeenCalledTimes(2)
+      expect(mockPostEvent.mock.calls[0][0]).toEqual({
```
small thing: I would recommend updating the doc string to reflect that we're tracking 2 events: bid and registration

dleve123 commented on src/Apps/Auction/Routes/__tests__/ConfirmBid.test.tsx:
```diff
@@ -438,8 +447,17 @@ describe("Routes/ConfirmBid", () => {
-      expect(mockPostEvent).toHaveBeenCalledTimes(1)
-      expect(mockPostEvent).toBeCalledWith({
+      expect(mockPostEvent).toHaveBeenCalledTimes(2)
+      expect(mockPostEvent.mock.calls[0][0]).toEqual({
```
oh this is interesting, so we're simulating the case where registration succeeds, but bidding fails. I'd recommend updating the docstring here to make this more clear.

dleve123 replied:
Also, are we testing the case when registration fails when attempting to bid+register?

yuki24 replied:
We can't track registration failed because it's all done [by Gravity](https://github.com/artsy/gravity/blob/5985f0f43078699d7ff81d963e08286f6d3db32f/app/api/v1/bidder_positions_endpoint.rb#L65-L72) in the single bidder position endpoint and errors from Gravity will be translated to a GraphQL error in Metaphysics.

dleve123 commented on src/Apps/Auction/Routes/ConfirmBid/index.tsx:
```diff
@@ -65,6 +67,9 @@ export const ConfirmBidRoute: React.FC<ConfirmBidProps> = props => {
     me as any
   )
 
+  let bidderId: string | null =
```
nice refactoring.

dleve123 commented on src/Apps/Auction/Routes/ConfirmBid/index.tsx:
```diff
@@ -54,6 +55,7 @@ const MAX_POLL_ATTEMPTS = 20
 
 export const ConfirmBidRoute: React.FC<ConfirmBidProps> = props => {
   let pollCount = 0
+  let registrationTracked = false
```
probably a newbie question - but why the need to maintain state here like this?

erikdstock replied:
Also wondering this (I think it has already been explained to me in some other context but why not use `useState`?

yuki24 replied:
> why the need to maintain state here like this?

The user would stay on the same page if the user wasn't registered, tried to bid and register but the user was outbid. This way the user will be able to place another bid without a lot of distraction (and also the simplest we could implement). We want to avoid calling the `registration submitted` event more than once in such a case.

> why not use `useState`?

We don't want the whole component to get re-rendered on every internal state update so components like `<OnSubmitValidationError>` won't fire the callback too many times and React won't run unnecessary VDOM diff-ing.

erikdstock commented on src/Apps/Auction/Components/RegistrationForm.tsx:
```diff
@@ -139,9 +132,9 @@ type TrackErrors = (errors: string[]) => void
+export const OnSubmitValidationError: React.FC<{
   cb: TrackErrors
-  formikProps: FormikProps<FormValues>
+  formikProps: Partial<FormikProps<FormValues>>
```
I wonder if this form should be moved up to a more common/shared directory- it seems like other forms may want to use it as well.

yuki24 replied:
Agreed - there are so many things shared across `Register` and `ConfirmBid` and they have imports from each other, which makes it harder to trace dependencies. Though that would probably be a number of incremental changes, so I wanted to focus on shipping first. cc @dblandin

yuki24 commented:
updated test descriptions and added one more test to cover the case explained [here](https://github.com/artsy/reaction/pull/3028#discussion_r349679423), merging.

artsyit commented:
<!-- GITHUB_RELEASE COMMENT: released -->
:rocket: PR was released in v22.7.0 :rocket:
